### PR TITLE
Fixes Item_modifier::modify() not setting curammo appropriately 

### PR DESCRIPTION
**Pull request by NickGrease**
This fixes a (currently) very specific problem when spawning `TOOL`'s with `"initial_charges": 0`, and used in item groups with charges set to a range, in this case `coffeemaker`'s. This leads to an issue when unloading the coffeemaker - giving the player a 'none' item instead of batteries.

From best I can tell, the sequence of events that leads to this is:

``` C++
Single_item_creator::create_single(...) {
  ...
  tmp = isd->create_single(birthday, rec); 
    item::item(...){
      ...
      charges = type->charges_default(); //creates coffeemaker from tools.json (initial_cahages=0)
      ...
      if( ammo_remaining()  && ( ammo_type() != "NULL" ) ) {
        set_curammo( default_ammo( ammo_type() ) ); //items with 0 charges never make it here
      }
    }
  modifier->modify(tmp); 
    Item_modifier::modify(...) //sets charges as per item_group definition [0,200]
    ...
    if( g != nullptr && ( ammo.get() != nullptr || ch > 0 ) ) { //coffeemaker is not a gun
      ...
      new_item.set_curammo( ammoid ); // never makes it here
    }
}
// curammo is never set
```

As far as I can tell this only affects the coffeemakers placed from hotel_tower.json, but it could affect a number of other `TOOL`'s if people decide to add charge ranges to more tools.

On that note, is the json member 'rand_charges' being phased in or out?

### CODE REVIEW

File: src/item.cpp [hunk 1]
```diff
@@ -137,7 +137,7 @@ item::item( const itype *type, int turn, int qty ) : type( type )
         active = goes_bad() && !rotten();
 
     } else if( type->is_tool() ) {
-        if( ammo_remaining() && ammo_type() != "NULL" ) {
+        if( ( ammo_remaining() >= 0 ) && ( ammo_type() != "NULL" ) ) {
```
Comment (mugling) [inline src/item.cpp]: `ammo_remaining() >= 0` will always evaluate true

**Comment by mugling** [comment]
> This leads to an issue when unloading the coffeemaker

It shouldn't be unloadable at `charges == 0`

**Comment by NickGrease** [comment]
> ammo_remaining() >= 0 will always evaluate true

I wasn't sure if charges could ever be (-)negative, just took the safe root.
> It shouldn't be unloadable at charges == 0.

It's not, the issue as best I could describe is in the comments above. Initial creation sets charges to 0 (as per tools.json initial_charges) but later in the creation process charges are set based on hotel_tower.json (`{ "item": "coffeemaker", "charges": [0, 200], "prob": 60 }`), which set the charges to something other than 0, but doesn't set curammo at the same time (or ever).

**Comment by mugling** [comment]
You probably need a call to `ammo_set` somewhere in `Item_modifier::modify` rather than just the call to `set_curammo`

**Comment by NickGrease** [comment]
```C++
if(ch != -1) {
        if( new_item.count_by_charges() || new_item.made_of( LIQUID ) ) {
            // food, ammo
            // count_by_charges requires that charges is at least 1. It makes no sense to
            // spawn a "water (0)" item.
            new_item.charges = std::max( 1l, ch );
        } else if(t != NULL) {
            new_item.charges = std::min(ch, t->max_charges);
====> would here work?
        } else if (g == nullptr){
            //not gun, food, ammo or tool. 
            new_item.charges = ch;
        }
    }
```

**Comment by NickGrease** [comment]
> You probably need a call to ammo_set somewhere in Item_modifier::modify rather than just the call to set_curammo

Just so I understand, I was only loosening the restriction to get to the already existing set_curammo, by  removing the check for charges > 0. Does setting curammo on tools with 0 charges have any side effect?

**Comment by mugling** [comment]
> Does setting curammo on tools with 0 charges have any side effect?

`curammo` should only be set if `charges > 0`
also `charges` is always `>= 0`

**Comment by NickGrease** [comment]
Got it, thanks - will update soon.

**Comment by NickGrease** [comment]
@mugling That should do it. I'm not sure if rng(0,xx) can return 0 or not, so the check for charges > 0 is added.

File: src/item_group.cpp [hunk 2]
```diff
@@ -191,7 +191,12 @@ void Item_modifier::modify(item &new_item) const
             // spawn a "water (0)" item.
             new_item.charges = std::max( 1l, ch );
         } else if(t != NULL) {
```
Comment (mugling) [inline src/item_group.cpp]: could be refactored as `new_item.is_tool()`

File: src/item_group.cpp [hunk 3]
```diff
@@ -191,7 +191,12 @@ void Item_modifier::modify(item &new_item) const
...
-            new_item.charges = std::min(ch, t->max_charges);
+            const auto a = new_item.ammo_type();
+            const auto c = std::min(ch, t->max_charges);
```
Comment (mugling) [inline src/item_group.cpp]: can be refactored as `new_item.ammo_capacity()`

Comment (mugling) [inline src/item_group.cpp]: try and avoid such short variable names (i appreciate the existing code sets a bad example). you could consider naming this variable `qty`

File: src/item_group.cpp [hunk 4]
```diff
@@ -191,7 +191,12 @@ void Item_modifier::modify(item &new_item) const
...
+            new_item.charges = c;
+            if( ( a != "NULL" ) && ( c > 0 ) ) {
```
Comment (mugling) [inline src/item_group.cpp]: operator precedence makes those extra parentheses redundant

File: src/item_group.cpp [hunk 5]
```diff
@@ -191,7 +191,12 @@ void Item_modifier::modify(item &new_item) const
             // spawn a "water (0)" item.
             new_item.charges = std::max( 1l, ch );
         } else if(t != NULL) {
-            new_item.charges = std::min(ch, t->max_charges);
+            const auto a = new_item.ammo_type();
```
Comment (mugling) [inline src/item_group.cpp]: you can probably just repeat the call to `ammo_type()` without defining a scoped variable. any sane compiler will generate identical code and there isn't a more descriptive variable name that you could use here

File: src/item.cpp [hunk 6]
```diff
@@ -137,7 +137,7 @@ item::item( const itype *type, int turn, int qty ) : type( type )
         active = goes_bad() && !rotten();
 
     } else if( type->is_tool() ) {
-        if( ammo_remaining() && ammo_type() != "NULL" ) {
+        if( ammo_remaining() && ( ammo_type() != "NULL" ) ) {
```
Comment (mugling) [inline src/item.cpp]: try and avoid making this formatting change

**Comment by mugling** [comment]
Otherwise looks like it should resolve the issue

**Comment by NickGrease** [comment]
@mugling Thanks for all your help, really appreciate it.

> could be refactored as new_item.is_tool()

So I can remove `const auto t = dynamic_cast<const it_tool *>(new_item.type)` since its only used there?

Current ( all comments taken)
```C++
} else if( new_item.is_tool() )  {
  const auto qty = std::min( ch, new_item.ammo_capacity() );
  new_item.charges = qty;
  if( new_item.ammo_type() != "NULL" && qty > 0 ) {
      new_item.ammo_set( new_item.ammo_type(), qty );
  }
}...
```

> try and avoid making this formatting change

Will do.

**Comment by mugling** [comment]
> So I can remove const auto t = dynamic_cast<const it_tool *>(new_item.type) since its only used there?

Yes, please do. It might be worth also dropping `g` in favor of a check for `is_gun`

> Current ( all comments taken)

Yes, much more legible

**Comment by NickGrease** [comment]
> Yes, please do. It might be worth also dropping g in favor of a check for is_gun

I'll make the changes and update soon.

**Comment by NickGrease** [comment]
I'll de-WiP with someone's approval. Sorry for all the commits.

**Pull request merged**